src/backend/catalog-builder/npm-downloads.lambda-shared.ts

```diff
@@ -0,0 +1,150 @@
+import type { Got } from 'got/dist/source/types';
+
+export enum NpmDownloadsPeriod {
+  /**
+   * Gets downloads for the last available day. In practice, this will usually
+   * be "yesterday" (in GMT) but if stats for that day have not yet landed, it
+   * will be the day before.
+   */
+  LAST_DAY = 'last-day',
+
+  /**
+   * Gets downloads for the last 7 available days.
+   */
+  LAST_WEEK = 'last-week',
+
+  /**
+   * Gets downloads for the last 30 available days.
+   */
+  LAST_MONTH = 'last-month'
+}
+
+export interface NpmDownloadsEntry {
+  readonly downloads: number;
+  readonly start: string;
+  readonly end: string;
+  readonly package: string;
+}
+
+export type NpmDownloadsOutput = { [key: string]: NpmDownloadsEntry };
+
+export interface NpmDownloadsOptions {
+  /**
+   * The period to query for package download counts.
+   * @default NpmDownloadsPeriod.LAST_WEEK
+   */
+  readonly period?: NpmDownloadsPeriod;
+
+  /**
+   * Throw an error when any package's download metrics are not available.
+   * @default true
+   */
+  readonly throwErrors?: boolean;
+}
+
+export class NpmDownloadsClient {
+  public static readonly NPM_DOWNLOADS_API_URL = 'https://api.npmjs.org/downloads/point';
+  public static readonly MAX_PACKAGES_PER_QUERY = 128; // this is a limitation from npm's API! do not change
+
+  private readonly got: Got;
+
+  constructor(gotService: Got) {
+    this.got = gotService;
+  }
+
+  private async getDownloadsRaw(
+    packages: string[],
```

Comment:
I like enforcing that we do not mutate our parameters where possible.

```suggestion
    packages: readonly string[],
```